File: sample.cs/TreasureIsland.Api/MartenSignalrConsumer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Glow.Core.Notifications;
using Glow.Invoices.Api.Test;
using Glow.NotificationsCore;
using Glow.Sample;
using Marten;
using Marten.Events;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;

namespace Glow.Core.Marten;

public class MartenSignalrConsumer : IMartenEventsConsumer
{
    private readonly IServiceProvider sp;

    public MartenSignalrConsumer(IServiceProvider sp)
    {
        this.sp = sp;
    }

    public async Task ConsumeAsync(IDocumentOperations documentOperations, IReadOnlyList<StreamAction> streamActions, CancellationToken ct)
    {
        using var scope = sp.CreateScope();
        var logger = scope.GetService<ILogger<MartenSignalrConsumer>>();

        var session = scope.GetService<IDocumentSession>();

        var games = await session.Query<Game>().ToListAsync();
        var current = games.FirstOrDefault();
        var svc = scope.GetService<IClientNotificationService>();

        if (current != null)
        {
            var playerIds = current.Units;
            var players = session.LoadMany<Unit>(playerIds);
            var dict = players.ToDictionary(v => v.Id, v => v);
            var notification = new CurrentGameState(current.Id, dict, current);

            await svc.PublishNotification(notification);
        }

        foreach (var actions in streamActions)
        {
            // logger.LogInformation("publish stream actions");
            foreach (var e in actions.Events)
            {
                // logger.LogInformation("publish event {@event}", e);

                logger.LogInformation("Publish event " + e.Data.GetType().Name);
                if (e.Data is IClientNotification clientNotification)
                {
                    await svc.PublishNotification(clientNotification);
                }
            }
        }
    }
}

File: samples/treasure-island.fs/MartenSignalrConsumer.fs
namespace TreasureIsland

open System
open FSharp.Control
open Glow.Core.MartenSubscriptions
open Glow.Core.Notifications
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Logging
open Glow.Core.Utils

type MartenSignalrConsumer(sp: IServiceProvider) =
  interface IMartenEventsConsumer with
    member this.ConsumeAsync(documentOperations, streamActions, ct) =
      task {
        use scope = sp.CreateScope()

        let logger =
          scope.Get<ILogger<MartenSignalrConsumer>>()

        let svc =
          scope.Get<IClientNotificationService>()

        let! result =
          streamActions
          |> AsyncSeq.ofSeq
          |> AsyncSeq.iterAsync (fun actions ->
            async {
              return!
                actions.Events
                |> AsyncSeq.ofSeq
                |> AsyncSeq.iterAsync (fun e ->
                  async {

                    if e.Data :? GameEvent then
                      logger.LogInformation("Publish game event " + e.Data.GetType().Name)

                      do!
                        svc.PublishNotification({ GameEventNotification.GameEvent = e.Data :?> GameEvent })
                        |> Async.AwaitTask
                    else
                      logger.LogInformation("Skip Publish game event " + e.Data.GetType().Name)
                    //
                    // if e.Data :? IClientNotification then
                    //   logger.LogInformation("Publish event " + e.Data.GetType().Name)
                    //
                    //   do!
                    //     svc.PublishNotification(e.Data :?> IClientNotification)
                    //     |> Async.AwaitTask
                    // else
                    //   logger.LogInformation("Publish event " + e.Data.GetType().Name)

                  })
            })
        // |> Async.StartAsTask

        // |> Seq.iter (fun actions ->
        //   actions.Events
        //   |> Seq.iter (fun e ->
        //     logger.LogInformation("Publish event " + e.Data.GetType().Name)
        //     if e.Data :? IClientNotificationService then
        //       do! svc.PublishNotification(e.Data :> IClientNotificationService)
        //
        //
        //   // if (e.Data is IClientNotification clientNotification)
        //   // {
        //   //     await svc.PublishNotification(clientNotification)
        //   // }
        //
        //   ))

        // let session =
        //   scope.GetService<IDocumentSession>()
        //
        // let! games = session.Query<Game>().ToListAsync()
        // let current = games.FirstOrDefault()
        //
        // if box current <> null then
        //
        //   let players =
        //     session.LoadMany<PlayerUnit>(
        //       current.PlayerUnitIds
        //       |> List.map (fun (PlayerUnitId id) -> id)
        //     )
        //
        //   let dict =
        //     players.ToDictionary(fun v -> v.Id)
        //
        //   let notification: CurrentGameState =
        //     { GameId = current.Id
        //       Units = dict
        //       Game = current }
        //
        //   do! svc.PublishNotification(notification)
      ()
      }
